### ISSUE TITLE
Test flake: chttp2_settings_timeout_test --gtest_filter=SettingsTimeout.Basic 

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#14214]
Flaky on TSAN with epoll1. Failing about 12% of builds.

We see an increased flakiness starting from ~Jan 17, but I didn't see a clear link between the flake and a single PR (and similar flakes have been seen in the past, just much less common).

https://sponge.corp.google.com/invocation?id=da55d59b-3f7e-42bf-8cbc-8c47c8a507c8
https://sponge.corp.google.com/invocation?id=5f69c2e2-2488-44d1-b5a6-0355fad6ebca
https://sponge.corp.google.com/invocation?id=62218b09-18a9-4517-b0af-6a2758195608
https://sponge.corp.google.com/invocation?id=7bddf59f-9c48-4a72-93f2-1dcb232e8404

**Comment by jtattermusch** [comment]
CC @dgquintas

**Comment by jtattermusch** [comment]
It's a data race according to TSAN.
```
[ RUN      ] SettingsTimeout.Basic
I0129 05:42:20.497496718   23495 settings_timeout_test.cc:223] starting server on localhost:29830
I0129 05:42:20.499005672   23495 settings_timeout_test.cc:227] starting client connect
I0129 05:42:20.499729577   23495 settings_timeout_test.cc:179] OnEventDone(): "No Error"
I0129 05:42:20.499770067   23495 settings_timeout_test.cc:231] starting client read
I0129 05:42:20.503351708   23495 settings_timeout_test.cc:179] OnEventDone(): "No Error"
I0129 05:42:20.503405808   23495 settings_timeout_test.cc:142] client read 63 bytes
I0129 05:42:21.503931710   23500 settings_timeout_test.cc:179] OnEventDone(): {"created":"@1517204541.503761893","description":"Socket closed","fd":8,"file":"src/core/lib/iomgr/tcp_posix.cc","file_line":421,"grpc_status":14,"target_address":"ipv6:[::1]:29830"}
==================
WARNING: ThreadSanitizer: data race (pid=23495)
  Write of size 1 at 0x7ffd81fbb018 by thread T3:
    #0 grpc_core::test::(anonymous namespace)::Client::EventState::OnEventDone(void*, grpc_error*) /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:182:20 (chttp2_settings_timeout_test+0x0000004b9d01)
    #1 exec_ctx_run(grpc_closure*, grpc_error*) /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.cc:38:3 (chttp2_settings_timeout_test+0x0000004cb7a9)
    #2 grpc_closure_run(char const*, int, grpc_closure*, grpc_error*) /var/local/git/grpc/./src/core/lib/iomgr/closure.h:258:5 (chttp2_settings_timeout_test+0x0000004d3153)
    #3 call_read_cb((anonymous namespace)::grpc_tcp*, grpc_error*) /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.cc:366:3 (chttp2_settings_timeout_test+0x0000004d9521)
    #4 tcp_do_read((anonymous namespace)::grpc_tcp*) /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.cc:419:5 (chttp2_settings_timeout_test+0x0000004da341)
    #5 tcp_continue_read((anonymous namespace)::grpc_tcp*) /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.cc:470:5 (chttp2_settings_timeout_test+0x0000004d96b8)
    #6 tcp_handle_read(void*, grpc_error*) /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.cc:487:5 (chttp2_settings_timeout_test+0x0000004d91b0)
    #7 exec_ctx_run(grpc_closure*, grpc_error*) /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.cc:38:3 (chttp2_settings_timeout_test+0x0000004cb7a9)
    #8 grpc_core::ExecCtx::Flush() /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.cc:125:9 (chttp2_settings_timeout_test+0x0000004cb533)
    #9 end_worker(grpc_pollset*, grpc_pollset_worker*, grpc_pollset_worker**) /var/local/git/grpc/src/core/lib/iomgr/ev_epoll1_linux.cc:916:7 (chttp2_settings_timeout_test+0x0000005117ee)
    #10 pollset_work(grpc_pollset*, grpc_pollset_worker**, long) /var/local/git/grpc/src/core/lib/iomgr/ev_epoll1_linux.cc:987:3 (chttp2_settings_timeout_test+0x00000050de31)
    #11 grpc_pollset_work(grpc_pollset*, grpc_pollset_worker**, long) /var/local/git/grpc/src/core/lib/iomgr/ev_posix.cc:225:10 (chttp2_settings_timeout_test+0x0000004ca4f1)
    #12 cq_next(grpc_completion_queue*, gpr_timespec, void*) /var/local/git/grpc/src/core/lib/surface/completion_queue.cc:932:23 (chttp2_settings_timeout_test+0x0000004eee59)
    #13 grpc_completion_queue_next /var/local/git/grpc/src/core/lib/surface/completion_queue.cc:1010:10 (chttp2_settings_timeout_test+0x0000004eda16)
    #14 grpc_core::test::(anonymous namespace)::ServerThread::Serve() /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:82:21 (chttp2_settings_timeout_test+0x0000004b6c49)
    #15 void std::_Mem_fn<void (grpc_core::test::(anonymous namespace)::ServerThread::*)()>::operator()<, void>(grpc_core::test::(anonymous namespace)::ServerThread*) const /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:569:11 (chttp2_settings_timeout_test+0x0000004b845b)
    #16 void std::_Bind<std::_Mem_fn<void (grpc_core::test::(anonymous namespace)::ServerThread::*)()> (grpc_core::test::(anonymous namespace)::ServerThread*)>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1263:11 (chttp2_settings_timeout_test+0x0000004b8391)
    #17 void std::_Bind<std::_Mem_fn<void (grpc_core::test::(anonymous namespace)::ServerThread::*)()> (grpc_core::test::(anonymous namespace)::ServerThread*)>::operator()<, void>() /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1321:11 (chttp2_settings_timeout_test+0x0000004b82c9)
    #18 void std::_Bind_simple<std::_Bind<std::_Mem_fn<void (grpc_core::test::(anonymous namespace)::ServerThread::*)()> (grpc_core::test::(anonymous namespace)::ServerThread*)> ()>::_M_invoke<>(std::_Index_tuple<>) /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1699:18 (chttp2_settings_timeout_test+0x0000004b8238)
    #19 std::_Bind_simple<std::_Bind<std::_Mem_fn<void (grpc_core::test::(anonymous namespace)::ServerThread::*)()> (grpc_core::test::(anonymous namespace)::ServerThread*)> ()>::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1688:16 (chttp2_settings_timeout_test+0x0000004b81e8)
    #20 std::thread::_Impl<std::_Bind_simple<std::_Bind<std::_Mem_fn<void (grpc_core::test::(anonymous namespace)::ServerThread::*)()> (grpc_core::test::(anonymous namespace)::ServerThread*)> ()> >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/thread:115:13 (chttp2_settings_timeout_test+0x0000004b7d5c)
    #21 <null> <null> (libstdc++.so.6+0x0000000b696f)

  Previous read of size 1 at 0x7ffd81fbb018 by main thread:
    #0 grpc_core::test::(anonymous namespace)::Client::EventState::done() const /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:172:32 (chttp2_settings_timeout_test+0x0000004b9d56)
    #1 grpc_core::test::(anonymous namespace)::Client::PollUntilDone(grpc_core::test::(anonymous namespace)::Client::EventState*, long) /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:200:31 (chttp2_settings_timeout_test+0x0000004b9b2e)
    #2 grpc_core::test::(anonymous namespace)::Client::ReadUntilError() /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:137:12 (chttp2_settings_timeout_test+0x0000004b6748)
    #3 grpc_core::test::(anonymous namespace)::SettingsTimeout_Basic_Test::TestBody() /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:232:108 (chttp2_settings_timeout_test+0x0000004b5945)
    #4 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2402:10 (chttp2_settings_timeout_test+0x00000090b5e2)
    #5 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2438:14 (chttp2_settings_timeout_test+0x0000008e688e)
    #6 testing::Test::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2474:5 (chttp2_settings_timeout_test+0x0000008c1f6b)
    #7 testing::TestInfo::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2656:5 (chttp2_settings_timeout_test+0x0000008c3161)
    #8 testing::TestCase::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2774:5 (chttp2_settings_timeout_test+0x0000008c3c00)
    #9 testing::internal::UnitTestImpl::RunAllTests() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:4649:11 (chttp2_settings_timeout_test+0x0000008ce53c)
    #10 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2402:10 (chttp2_settings_timeout_test+0x0000009126d2)
    #11 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2438:14 (chttp2_settings_timeout_test+0x0000008ea2b4)
    #12 testing::UnitTest::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:4257:10 (chttp2_settings_timeout_test+0x0000008ce056)
    #13 RUN_ALL_TESTS() /var/local/git/grpc/third_party/googletest/googletest/include/gtest/gtest.h:2233:10 (chttp2_settings_timeout_test+0x0000004b9eed)
    #14 main /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:251:16 (chttp2_settings_timeout_test+0x0000004b563a)

  Location is stack of main thread.

  Thread T3 (tid=23500, running) created by main thread at:
    #0 pthread_create /llvm/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:954 (chttp2_settings_timeout_test+0x000000452373)
    #1 std::thread::_M_start_thread(std::shared_ptr<std::thread::_Impl_base>) <null> (libstdc++.so.6+0x0000000b6a90)
    #2 grpc_core::test::(anonymous namespace)::ServerThread::Start() /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:62:23 (chttp2_settings_timeout_test+0x0000004b5e73)
    #3 grpc_core::test::(anonymous namespace)::SettingsTimeout_Basic_Test::TestBody() /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:225:3 (chttp2_settings_timeout_test+0x0000004b58dc)
    #4 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2402:10 (chttp2_settings_timeout_test+0x00000090b5e2)
    #5 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2438:14 (chttp2_settings_timeout_test+0x0000008e688e)
    #6 testing::Test::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2474:5 (chttp2_settings_timeout_test+0x0000008c1f6b)
    #7 testing::TestInfo::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2656:5 (chttp2_settings_timeout_test+0x0000008c3161)
    #8 testing::TestCase::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2774:5 (chttp2_settings_timeout_test+0x0000008c3c00)
    #9 testing::internal::UnitTestImpl::RunAllTests() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:4649:11 (chttp2_settings_timeout_test+0x0000008ce53c)
    #10 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2402:10 (chttp2_settings_timeout_test+0x0000009126d2)
    #11 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:2438:14 (chttp2_settings_timeout_test+0x0000008ea2b4)
    #12 testing::UnitTest::Run() /var/local/git/grpc/third_party/googletest/googletest/src/gtest.cc:4257:10 (chttp2_settings_timeout_test+0x0000008ce056)
    #13 RUN_ALL_TESTS() /var/local/git/grpc/third_party/googletest/googletest/include/gtest/gtest.h:2233:10 (chttp2_settings_timeout_test+0x0000004b9eed)
    #14 main /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:251:16 (chttp2_settings_timeout_test+0x0000004b563a)

SUMMARY: ThreadSanitizer: data race /var/local/git/grpc/test/core/transport/chttp2/settings_timeout_test.cc:182:20 in grpc_core::test::(anonymous namespace)::Client::EventState::OnEventDone(void*, grpc_error*)
==================
Command exited with non-zero status 66
real 1.22
user 0.00
sys 0.06

2018-01-29 05:42:21,666 FAILED: bins/tsan/chttp2_settings_timeout_test --gtest_filter=SettingsTimeout.Basic  GRPC_POLL_STRATEGY=epoll1 [ret=66, pid=23494, time=1.2sec]
```

**Comment by dgquintas** [comment]
Seems to have been failing on and off for a while: https://screenshot.googleplex.com/kQUODhaQKhc

**Comment by dgquintas** [comment]
@yang-g could you have a quick look?

**Comment by jtattermusch** [comment]
Fixed by #14233.
